### ISSUE TITLE
future Connection and ConnectionManager breaking changes

### ISSUE THREAD

**Issue by pleerock** [typeorm/typeorm#7428]
## Connection confusion

Right now what we call a `Connection` isn't technically a connection. When we talk about real connection we usually mean an established connection to a database, however in TypeORM `Connection` isn't such. What Connection in TypeORM is - it's just a place where we store information about connection for future connections and hold a connections pool and some functions managing this connection. In TypeORM real database connection is called `QueryRunner`.

In future version we are planning to remove this confusion by renaming `Connection` into `DataSource` and `QueryRunner` into `Connection`.

## New way of connection creation

Right now we connection creation is established this way:

```ts
const connection = await createConnection({ ... })
```

Because you can't use top-level await (at least yet), you most probably will write a code following way:

```ts
createConnection({ ... }).then(connection => {
 ...
})
```

But this way you can't (without additional dirty code) export connection, so you can use it everywhere else.
That's why we have a `ConnectionManager` to collect all created connections, we have to "name" connections and we have bunch of exported functions like `getConnection(name)`, `getConnectionManager()`, etc.

In future version we'll remove all this redundant complexity by applying a new strategy for connection creation.

With a new default strategy you can create connection and write it into exported variable:

```ts
export const MyConnection = Connection.create({
   // ... connection options ...
})
```

Then you can use `MyConnection.manager`, `MyConnection.repository(User)` anywhere in the app.
The only caveat is that you need to call `MyConnection.init()` (new name for `.connect()`) somewhere in the bootstrapping area of the application.
 
And yes, since we just said we'll rename `Connection` to `DataSource`, next syntax will be:

```ts
export const MyDataSource = DataSource.create({
   // ... data source options ...
})
```

## `ConnectionManager` deprecation

`ConnectionManager` isn't necessary anymore because of the new default connection creation strategy. It will be removed.

## Global functions

All following global functions will be deprecated and removed:

- getConnectionOptions - options are defined in the connection creation
- getConnectionManager - `ConnectionManager` will be removed
- createConnection - `DataSource.create` should be used instead
- createConnections - multiple `DataSource.create` calls should be used instead
- getConnection - exported connection is used to get a connection
- getManager - exported connection properties should be used instead
- getMongoManager - exported connection properties should be used instead
- getSqljsManager - exported connection properties should be used instead
- getRepository - exported connection properties should be used instead
- getTreeRepository - exported connection properties should be used instead
- getCustomRepository - exported connection properties should be used instead
- getMongoRepository - exported connection properties should be used instead
- createQueryBuilder - exported connection properties should be used instead

**Comment by PodaruDragos** [comment]
@pleerock do you have an estimate when Milestone 2.0.0 will became actively worked on ?

**Comment by MusiCode1** [comment]
Hello, I'm new here.
All tutorials on the official pages instruct `createConnection` for connection.
If this code is no longer in use, what does the get started example look like for beginners?

**Comment by CMTV** [comment]
> Hello, I'm new here. All tutorials on the official pages instruct `createConnection` for connection. If this code is no longer in use, what does the get started example look like for beginners?

Create a `DataSource` instance with connection options and then call `initialize()` method:

```ts
let dataSource = new DataSource({
    type: 'better-sqlite3',
    database: 'data.db'
});

let connection = await dataSource.initialize();

connection.manager./* your stuff...*/
```

**Comment by pleerock** [comment]
@MusiCode1 please follow @CMTV answer. We'll update the official documentation very soon. Feel free to contribute.

**Comment by pleerock** [comment]
Closing as it was implemented in #8616
